components/prism-promql.js

```diff
@@ -0,0 +1,148 @@
+	var aggregations = [
+		'sum',
+		'min',
+		'max',
+		'avg',
+		'group',
+		'stddev',
+		'stdvar',
+		'count',
+		'count_values',
+		'bottomk',
+		'topk',
+		'quantile'
+	];
+
+	// PromQL functions
+	// (https://prometheus.io/docs/prometheus/latest/querying/functions/)
+	var functions = [
+		'abs',
+		'absent',
+		'ceil',
+		'changes',
+		'clamp_max',
+		'clamp_min',
+		'day_of_month',
+		'day_of_week',
+		'days_in_month',
+		'delta',
+		'deriv',
+		'exp',
+		'floor',
+		'histogram_quantile',
+		'holt_winters',
+		'hour',
+		'idelta',
+		'increase',
+		'irate',
+		'label_join',
+		'label_replace',
+		'ln',
+		'log2',
+		'log10',
+		'minute',
+		'month',
+		'predict_linear',
+		'rate',
+		'resets',
+		'round',
+		'scalar',
+		'sort',
+		'sort_desc',
+		'sqrt',
+		'time',
+		'timestamp',
+		'vector',
+		'year'
+	];
+
+	// PromQL specific functions: Aggregations over time
+	// (https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time)
+	aggregations.forEach(function (agg) {
+		functions.push(agg + '_over_time');
+	});
+
+	// PromQL vector matching + the by and without clauses
+	// (https://prometheus.io/docs/prometheus/latest/querying/operators/#vector-matching)
+	var vectorMatching = [
+		'on',
+		'ignoring',
+		'group_right',
+		'group_left',
+		'by',
+		'without',
+	];
+
+	// PromQL offset modifier
+	// (https://prometheus.io/docs/prometheus/latest/querying/basics/#offset-modifier)
+	var offsetModifier = ['offset'];
+
+	var keywords = aggregations.concat(vectorMatching, offsetModifier);
+
+	Prism.languages.promql = {
+		'comment': {
+			pattern: /(^[ \t]*)#.*/m,
+			lookbehind: true
+		},
+		'vector-match': {
+			// Match the comma-separated label lists inside vector matching:
+			pattern: new RegExp('((?:' + vectorMatching.join('|') + ')\\s*)\\([^)]*\\)'),
+			lookbehind: true,
+			inside: {
+				'label-key': {
+					pattern: /\b[^,]*\b/,
+					alias: 'attr-name',
+				},
+				'punctuation': /[(),]/
+			},
+		},
+		'context-labels': {
+			pattern: /\{[^{}]*\}/,
+			inside: {
+				'label-key': {
+					pattern: /\b[a-z_]\w*(?=\s*(?:=~?|![=~]))/,
+					alias: 'attr-name',
+				},
+				'label-value': {
+					pattern: /(["'`])(?:\\[\s\S]|(?!\1)[^\\])*\1/,
+					greedy: true,
+					alias: 'attr-value',
+				},
+				'punctuation': /\{|\}|=~?|![=~]|,/,
+			},
+		},
+		'context-range': [
+			{
+				pattern: /\[[\w\s:]+\]/, // [1m]
+				inside: {
+					'punctuation': /\[|\]|:/,
+					'range-duration': {
+						pattern: /\b(?:\d+(?:[smhdwy]|ms))+\b/i,
+						alias: 'number',
+					},
+				},
+			},
+			{
+				pattern: /(\boffset\s+)\w+/, // offset 1m
+				lookbehind: true,
+				inside: {
+					'range-duration': {
+						pattern: /\b(?:\d+(?:[smhdwy]|ms))+\b/i,
+						alias: 'number',
+					},
+				},
+			},
+		],
+		'function': new RegExp('\\b(?:' + functions.join('|') + ')(?=\\s*\\()', 'i'),
+		'keyword': new RegExp('\\b(?:' + keywords.join('|') + ')\\b', 'i'),
```

Comment:
After all keywords are matched, could we say that all functions are of the form `\b[a-zA-Z_]\w*(?=\s*\()`?

Reply:
I think we can, yes. But I do think there is significant value in listing the functions explicitly, as it allows one to instantly see by its color when they have mistyped a function name.

Reply:
The main problem with the listing approach is that we have to maintain the list. New versions might add/remove functions and we'd have to follow suit. Lists also increase the file size of the language definition. The only upside I can see is the typo detection you mentioned but Prism isn't an editor.

I also want to point out that functions are [parsed generically by PromQL's parser](https://github.com/prometheus/prometheus/blob/cda52234eb5b2db1a7cc33cba030354da658e131/promql/parser/generated_parser.y#L329). I think that we should do the same.

Reply:
Fair enough. I can monkey-patch it back in for our own project, so I won't object to using a generic approach upstream :)

Reply:
Updated.